### DISCUSSION TITLE
Registering namespaced livewire components like laravel blade

### DISCUSSION THREAD

**Discussion by walirazzaq** [livewire/livewire#4496] [3. Feature ideas]
Currently manual registration of livewire components for packages or from other location than the defined `namespace` in config is to register components one by one like so `Livewire::component('alias',Component::class)`. 
Laravel blade has another way of registering `namespaced` components like 
```php
Blade::componentNamespace('Nightshade\\Views\\Components', 'nightshade');
```
which then can be used like 
```blade
<x-nightshade::button />
```
I was wondering if we can do something similar to register livewire components 
```php
Livewire::componentNamespace('Acme\\Livewire','acme');
```
and the use them like
```blade
<livewire:acme::contact-form />
@livewire('acme::contact-form')
```
This can be really useful when commonly used components are extracted as package or when module based applications

**Comment by joshhanley** [comment]
@walirazzaq yeah agreed that could be useful! Even just being able to point to a directory and say register all components within it instead of having to register each component manually would be helpful.

**Comment by Wulfheart** [comment]
Related to https://github.com/livewire/livewire/discussions/1984

**Comment by RealMrHex** [comment]
@walirazzaq @joshhanley I have written a module to work with Livewire in the nwidart/laravel-modules structure and it does exactly what you want.
Honestly, I didn't have time to write the tests and since everything was working as expected I put off writing the tests 😶‍🌫️ 
If needed, I can share the module with you 😀

**Comment by RealMrHex** [comment]
@walirazzaq
This is what I used in the ServiceProvider of the module, Maybe can helps you :)
![image](https://user-images.githubusercontent.com/62525184/213296815-2f834cde-837e-4cfb-8414-e058762d7ec0.png)
![image](https://user-images.githubusercontent.com/62525184/213296942-782268f0-ad81-4424-a88f-0d87e2151e2a.png)
![image](https://user-images.githubusercontent.com/62525184/213296984-99dd4523-d487-4647-9f7e-374f166644a3.png)
**And here is the result**
![image](https://user-images.githubusercontent.com/62525184/213298272-83ae6292-5c71-4033-aff1-4615456e851a.png)

**Reply by Wulfheart** [reply]
Could you please paste the code itself?

**Reply by RealMrHex** [reply]
@Wulfheart Sure, here u go
[WireServiceProvider.php](https://gist.github.com/RealMrHex/2adff5d47ed1e4c0d5f0edaa4c99aaf2)

**Reply by joshhanley** [reply]
@RealMrHex yeah I do something similar here https://github.com/joshhanley/livewire-dusk-extension/blob/bdea5635c16dda3ab0d9c880f3a8fd7530379fdf/src/LivewireDuskExtensionServiceProvider.php#L26-L32

**Comment by RalphMRivera** [comment]
This is the way I handle this in my DDD project:

Let's assume, I have a namespace called `Dune` with a domain called `Atreides` that has dedicated folders named `Livewire` and `Views`. Let's also assume that I don't want any code or views in the default `App` namespace at `\app`.

## First, add your namespace to your composer file:

```
  "autoload" : {
    "psr-4" : {
      "App\\" : "app/",
      "Database\\Factories\\" : "database/factories/",
      "Database\\Seeders\\" : "database/seeders/",
      "Dune\\" : "Dune/"
    }
  },
```

## Create your component

```
\Dune\Atreides\Livewire\FearIsTheMindKillerLivewaire.php
```

## Create your View

```
\Dune\Atreides\View\fear-is-the-mind-killer.blade.php
```


## Update the provider file or a custom provider file

```
use Dune\Atreides\Livewire\FearIsTheMindKillerLivewaire

$this->loadViewsFrom(base_path() . '/Dune, 'dune');
Livewire::component('fear-is-the-mind-killer', FearIsTheMindKillerLivewire::class);
```

## Your component will look something like this

```php
namespace Dune\Atreides\Livewire

use Illuminate\Contracts\View\View;
use Livewire\Component;

class FearIsTheMindKillerLivewire extends Component
{
    public function render(): view
    {
        return view('dune::Atreides.View.fear-is-the-mind-killer');
    }
}
```

## In Blade, you would reference the Livewire component like this:

```
<div class="prose max-w-none">
  <h1>Fear Is The Mind Kille</h1>
  <livewire:fear-is-the-mind-killer />
</div>
```

## Misc

- The biggest downside is that if you have a lot of Livewire components, you need to register each. I don't find that a problem, but some might.
- My needs are fairly basic; This works for me , but there may be more advanced functionality that does not work.

**Reply by nickpoulos** [reply]
Thanks for providing your solution for this.  

It is frankly a little absurd the hoops you have to jump through for the basic feature of "let me choose where to put my classes". 🙄

**Comment by joserick** [comment]
I have had the same problem so I have created a package which allows you to load components from different namespaces, note that this is only for livewire 3.

[Laravel Livewire Discover - https://github.com/joserick/laravel-livewire-discover](https://github.com/joserick/laravel-livewire-discover)

``` php
public function boot(): void
{
	// Load multiples namespace for Livewire components.
	Livewire::discover('my-components', 'Namespace\\Livewire');
	Livewire::discover('new-components', 'User\\Repository\\Livewire');
	...
}
```
``` html
<livewire:my-components.devices />
<livewire:new-components.devices-table />
```

**Reply by calebporzio** [reply]
Very nice!

**Reply by juliomotol** [reply]
Would be nice if this was implemented to Livewire itself, I hope you'd consider @calebporzio.
